[BugFix] Let submitted tasks without execution can be awared in starrocks::LakeServiceImpl to set a correct response and status

srlch opened the pull request:
## Why I'm doing:
Currently, some task submitted by starrocks::LakeServiceImpl can not aware whether the task is executed or not,
which will cause the response or status to be set incorrectly.

## What I'm doing:
Change AutoCleanRunnable to CancellableRunnable, which has a canceller to be called when threadpool is shutdown
to handle the correct response and status. Remove the cleaner and caller should do the proper clean-up work
in run() or when the task submitted failed.

Fixes #issue

## What type of PR is this:

- [x] BugFix
- [ ] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function
- [ ] This is a backport pr

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.5
  - [x] 3.4
  - [x] 3.3

xiangguangyxg commented on be/src/service/service_be/lake_service.cpp:
```diff
@@ -178,9 +178,12 @@ void LakeServiceImpl::publish_version(::google::protobuf::RpcController* control
     for (auto tablet_id : request->tablet_ids()) {
         auto task = std::make_shared<AutoCleanRunnable>(
                 [&, tablet_id] {
+                    DeferOp defer([&]() { no_execution_tablets.erase(tablet_id); });
```
need lock ?

srlch replied:
Fixed

xiangguangyxg commented on be/src/util/threadpool.h:
```diff
@@ -79,15 +79,22 @@ class Runnable {
 public:
-    AutoCleanRunnable(std::function<void()> runner, std::function<void()> cleaner)
-            : _runnable(std::move(runner)), _cleaner(std::move(cleaner)) {}
-    virtual ~AutoCleanRunnable() { _cleaner(); }
+    AutoCleanRunnable(std::function<void()> runner, std::function<void(Status)> cleaner)
+            : _runnable(std::move(runner)), _cleaner(std::move(cleaner)), _st(Status::Unknown("Not yet executed")) {}
+    virtual ~AutoCleanRunnable() { _cleaner(_st); }
 
-    virtual void run() override { _runnable(); }
+    virtual void run() override {
+        set_status(Status::OK());
+        _runnable();
+    }
+
+    void set_status(Status st) { _st = st; }
```
Would it be simpler to execute only one of the runner or cleaner ?

srlch replied:
Could you give some detail about how to just run one of them? It seems that we always need two functions here.

xiangguangyxg replied:
one is for normal execution, another is for abort execution. Use an atomic bool flag to garanteen only one will be executed.

srlch replied:
Updated

xiangguangyxg commented on be/src/util/threadpool.cpp:
```diff
@@ -707,9 +730,21 @@ Status ConcurrencyLimitedThreadPoolToken::submit(std::shared_ptr<Runnable> task,
-    auto token_task =
-            std::make_shared<AutoCleanRunnable>([t = std::move(task)] { t->run(); }, [sem = _sem] { sem->release(); });
-    return _pool->submit(std::move(token_task));
+    auto token_task = std::make_shared<CancellableRunnable>(
+            [t = task, sem = _sem] {
+                t->run();
+                sem->release();
+            },
+            [t = task, sem = _sem] {
+                t->cancel();
+                sem->release();
+            });
```
Using a specific class implemented runable here could avoid coping task.

kevincai commented on be/src/util/threadpool.h:
```diff
@@ -68,26 +68,30 @@ class Runnable {
+// correct clean-up work in the corrsponding case:
+// 1. Schedule by the threadpool and execute run().
+// 2. Cancelled by the threadpool when it has been shutdown(), runnable will be called cancel().
+// 3. Submit runnable task failed into the threadpool.
```
```suggestion
// 3. Submit runnable task failed into the threadpool. Neither `run()` nor `cancel()` will be invoked.
```

kevincai commented on be/src/util/threadpool.cpp:
```diff
@@ -129,6 +130,15 @@ void ThreadPoolToken::shutdown() {
     PriorityQueue<ThreadPool::NUM_PRIORITY, ThreadPool::Task> to_release;
+    DeferOp defer([&]() {
+        while (!to_release.empty()) {
+            (to_release.front().runnable)->cancel();
```
make it exception safe, so every task will have chance to invoke `cancel()`

kevincai commented on be/src/util/threadpool.h:
```diff
@@ -68,26 +68,30 @@ class Runnable {
+// 1. Schedule by the threadpool and execute run().
+// 2. Cancelled by the threadpool when it has been shutdown(), runnable will be called cancel().
+// 3. Submit runnable task failed into the threadpool.
+class CancellableRunnable : public Runnable {
```
need to fix the comments of ThreadPool::submit and ThreadPool::submit_func

kevincai commented on be/src/service/service_be/lake_service.cpp:
```diff
@@ -1168,12 +1273,22 @@ void LakeServiceImpl::vacuum_full(::google::protobuf::RpcController* controller,
+            },
+            [&] {
+                Status st = Status::Cancelled("full vacuum task has been cancelled");
+                LOG(WARNING) << st;
```
this logline has no useful information, shall print at least the tablet id, or remove this logline

srlch replied:
Fixed

xiangguangyxg commented on be/src/util/threadpool.h:
```diff
@@ -68,26 +68,30 @@ class Runnable {
-            : _runnable(std::move(runner)), _cleaner(std::move(cleaner)) {}
-    virtual ~AutoCleanRunnable() { _cleaner(); }
+    CancellableRunnable(std::function<void()> runner, std::function<void()> canceller)
+            : _runnable(std::move(runner)), _canceller(std::move(canceller)) {}
+    virtual ~CancellableRunnable() = default;
 
     virtual void run() override { _runnable(); }
 
+    virtual void cancel() override { _canceller(); }
+
 protected:
     std::function<void()> _runnable;
-    std::function<void()> _cleaner;
+    std::function<void()> _canceller;
```
Will it be simpler if there are 3 tasks? runnable, canceller and cleaner, then you only need to add the canceller to handle the cancellation, runnable and cleaner are keep unchanged.

xiangguangyxg replied:
Yes but without the _cleaner, there are so many explicit `count_down()` added in many places, they should be eliminated, or the code is hard to maintain.

xiangguangyxg replied:
Can the explicit `count_down()` be eliminated by using `DeferOP`?

srlch committed:
[BugFix] Let submitted tasks without execution can be awared in starrocks::LakeServiceImpl to set a correct response and status

## Why I'm doing:
Currently, some task submitted by starrocks::LakeServiceImpl can not aware whether the task is executed or not,
which will cause the response or status to be set incorrectly.

## What I'm doing:
Change AutoCleanRunnable to CancellableRunnable, which has a canceller to be called when threadpool is shutdown
to handle the correct response and status. Remove the cleaner and caller should do the proper clean-up work
in run() or when the task submitted failed.

Signed-off-by: srlch <linzichao@starrocks.com>

srlch committed:
fix

Signed-off-by: srlch <linzichao@starrocks.com>
